Source/EasyNetQ/Consumer/ConsumerDispatcher.cs

```diff
@@ -8,53 +8,68 @@ namespace EasyNetQ.Consumer
     public class ConsumerDispatcher : IConsumerDispatcher
     {
         private readonly ILog logger = LogProvider.For<ConsumerDispatcher>();
-        private readonly BlockingCollection<Action> queue;
-        private bool disposed;
+        private readonly CancellationTokenSource cancellation = new CancellationTokenSource();
+        private readonly SemaphoreSlim actionsAvailable = new SemaphoreSlim(0);
+        private readonly ConcurrentQueue<Action> durableActions = new ConcurrentQueue<Action>();
+        private readonly ConcurrentQueue<Action> transientActions = new ConcurrentQueue<Action>();
 
         public ConsumerDispatcher(ConnectionConfiguration configuration)
         {
             Preconditions.CheckNotNull(configuration, "configuration");
 
-            queue = new BlockingCollection<Action>();
-
             var thread = new Thread(_ =>
             {
-                while (!disposed && queue.TryTake(out var action, -1))
-                {
+                while (!cancellation.IsCancellationRequested || !durableActions.IsEmpty || !transientActions.IsEmpty)
                     try
                     {
-                        action();
+                        if (durableActions.TryDequeue(out var action) || transientActions.TryDequeue(out action))
+                            action();
+                        else
+                            actionsAvailable.Wait(cancellation.Token);
```

Comment:
nice implementation :) would an await instead wait be more effiecient ? :) "await actionsAvailable.WaitAsync(cancellation.Token)"

Reply:
Nope, because EasyNetQ uses dedicated thread here. Also using async code here could leed to hijacking of this thread, so let's don't change sync to async here.

Reply:
That makes sense :)

Reply:
At this dedicated thread we need exactly `Wait`, it's OK, `await` does not fit at all.